### ISSUE TITLE
False positive when updating calendar with DAVx5

### ISSUE THREAD

**Issue by ti-guru** [coreruleset/nextcloud-rule-exclusions-plugin#122]
Hi,

I've set up the nextcloud rule exclusions with caddy and coraza running in nested Docker containers in LXD. When creating a calendar event on my android phone and syncing it to the server through [DAVx5](https://play.google.com/store/apps/details?id=at.bitfire.davdroid&pli=1) I get the following error logged to console:
```
caddy-1  | {"level":"error","ts":1747254609.1309803,"logger":"http.handlers.waf","msg":"[client \"127.0.0.1\"] Coraza: Warning. HTTP Header Injection Attack via payload (CR/LF detected) [file \"@owasp_crs/REQUEST-921-PROTOCOL-ATTACK.conf\"] [line \"4777\"] [id \"921150\"] [rev \"\"] [msg \"HTTP Header Injection Attack via payload (CR/LF detected)\"] [data \"Matched Data: \\r found within ARGS_NAMES:BEGIN:VCALENDAR\\r\\nVERSION:2.0\\r\\nPRODID:DAVx5/4.4.10-gplay ical4j/3.2.19\\r\\nBEGIN:VEVENT\\r\\nDTSTAMP:20250514T203008Z\\r\\nUID:<longUID string>\\r\\nSUMMARY:Test\\r\\nDTSTART;TZID: BEGIN:VCALENDAR\\r\\nVERSION:2.0\\r\\nPRODID:DAVx5/4.4.10-gplay ica\"] [severity \"critical\"] [ver \"OWASP_CRS/4.7.0\"] [maturity \"0\"] [accuracy \"0\"] [tag \"application-multi\"] [tag \"language-multi\"] [tag \"platform-multi\"] [tag \"attack-protocol\"] [tag \"paranoia-level/1\"] [tag \"OWASP_CRS\"] [tag \"capec/1000/210/272/220/33\"] [hostname \"\"] [uri \"/remote.php/dav/calendars/<username>/<calendar name>/<calendar event id>.ics\"] [unique_id \"VnlVOaXIbllaCyMR\"]"}
caddy-1  | {"level":"error","ts":1747254609.1400568,"logger":"http.handlers.waf","msg":"[client \"127.0.0.1\"] Coraza: Access denied (phase 2). Inbound Anomaly Score Exceeded (Total Score: 5) [file \"@owasp_crs/REQUEST-949-BLOCKING-EVALUATION.conf\"] [line \"9194\"] [id \"949110\"] [rev \"\"] [msg \"Inbound Anomaly Score Exceeded (Total Score: 5)\"] [data \"\"] [severity \"emergency\"] [ver \"OWASP_CRS/4.7.0\"] [maturity \"0\"] [accuracy \"0\"] [tag \"anomaly-evaluation\"] [tag \"OWASP_CRS\"] [hostname \"\"] [uri \"/remote.php/dav/calendars/<username>/<calendar name>/<calendar event id>.ics\"] [unique_id \"VnlVOaXIbllaCyMR\"]"}
```
If I add SecRuleRemoveById 921150 in my Caddyfile the error disappears and the calendar event is synced successfully. Is it possible to fix this in the exclusion list without globally removing 921150? This has also occurred with rule id 921120. 

Software: 
Nextcloud-rule-exclusions-plugin 1.4.0
[Coraza-caddy 2.0.0](https://github.com/corazawaf/coraza-caddy)
[Nextcloud AIO 10.14.0 (Hub 9, 30.0.10)](https://github.com/nextcloud/all-in-one)

<details>
  <summary>Caddyfile in reverse proxy</summary>

  ```
  {
# Turn on debug mode
       debug

# See https://github.com/corazawaf/coraza-caddy
        order coraza_waf first

# See https://github.com/mastercactapus/caddy2-proxyprotocol
        servers {
                listener_wrappers {
                        proxy_protocol {
                                #timeout 5s
                                #allow <cidrs...>
                        }
                        tls
                }
        }

}

cloud.example.com:443 {

        handle /push/ws {
                @websockets {
                        header_regexp Connection Upgrade
                        header Upgrade websocket
                }
                reverse_proxy @websockets nextcloud.lxd:11000
        }

        handle /cool/* {
                @websockets {
                        header_regexp Connection Upgrade
                        header Upgrade websocket
                }
                reverse_proxy @websockets nextcloud.lxd:11000
        }


        handle * {

                coraza_waf {
                        load_owasp_crs
                        directives `
                                Include /nextcloud-rule-exclusions-plugin/plugins/*.conf
                                Include @coraza.conf-recommended
                                Include @crs-setup.conf.example
                                Include @owasp_crs/*.conf

                                #SecRuleRemoveById 921150
                                SecRequestBodyLimit 1073741824

                                SecRuleEngine On
                        `
                }

                reverse_proxy nextcloud.lxd:11000

        }

        tls {
               dns cloudflare <TOP SECRET KEY>
       }
}
  ```
</details>

**Comment by ti-guru** [comment]
Thanks for reaching back! Simply modifying exclusion rule [9508338](https://github.com/coreruleset/nextcloud-rule-exclusions-plugin/blob/5397c3c441888dd49a29ade017ca5aaf1cff00e3/plugins/nextcloud-rule-exclusions-before.conf#L732) solves my issue:

```
\# Editing calendar tasks notes.
SecRule REQUEST_FILENAME "@rx /remote\.php/dav/calendars/(?:[^/]+/){2}[^/]+\.ics$" \
    "id:9508338,\
    phase:1,\
    pass,\
    t:none,\
    nolog,\
    ver:'nextcloud-rule-exclusions-plugin/1.4.0',\
    ctl:ruleRemoveById=920530,\
    ctl:ruleRemoveById=921150"
```

Is this something that should go into the config? Looking at the [crs-setup.conf](https://github.com/corazawaf/coraza-coreruleset/blob/9b73b5e90a09613c7535d391a15cc9cb08c05a8d/rules/%40crs-setup.conf.example#L217) it does not appear to be enabled.

**Comment by ti-guru** [comment]
Also, before posting another issue. I'm also getting lots of false positives when uploading images to my instance, such as:

```
caddy-1  | {"level":"error","ts":1747345651.6303551,"logger":"http.handlers.waf","msg":"[client \"127.0.0.1\"] Coraza: Warning. Invalid character in request (null character) [file \"@owasp_crs/REQUEST-920-PROTOCOL-ENFORCEMENT.conf\"] [line \"3813\"] [id \"920270\"] [rev \"\"] [msg \"Invalid character in request (null character)\"] [data \"ARGS:HB\\x82\\x81\\xf8\\xfc:\\xdce\\\"\\xea=\\xfe\\x91r\\x90\\x8c\\x00\\xf7\\xe9qd6Q\\xef\\xba\\xc1\\x18E\\x87Ϻ\\xd3kb\\xf3\\a7\\x1dρXsV\\xae\\xfa6A\\xda\\xc7X1\\x11K\\xb2\\xf0\\x807\\x91\\xf2\\x0f[\\xb8B\\x88ׇ\\x90|\\xcdڴ\\x8c`\\xf4\\xdf\\xe9 \\xa9\\xbb\\x02\\x13\"] [severity \"critical\"] [ver \"OWASP_CRS/4.7.0\"] [maturity \"0\"] [accuracy \"0\"] [tag \"application-multi\"] [tag \"language-multi\"] [tag \"platform-multi\"] [tag \"attack-protocol\"] [tag \"paranoia-level/1\"] [tag \"OWASP_CRS\"] [tag \"capec/1000/210/272\"] [hostname \"\"] [uri \"/remote.php/dav/files/<username>/example.jpg\"] [unique_id \"eNDUzmcDreTximUX\"][msg_match_1 \"Invalid character in request (null character)\"] [data_match_1 \"ARGS:\\xe0\\xdbMz\\x81\\xfa=\\x8c\\xaa\\x00\\xc8H\\xd6\\xdc;zK\\xa8\\xf7ۓ\\xbf(t\\xe6\\xee\\xdac\\xf4\\xf4I\\x924\\x8c \\x11a{\\x0e1\\xc6\\x1f\\x90z\\xeeI\\xe0\\xf5\\x02\\xa2D\\xca1\\xa0!A\\x02D) \\xe0\\x1f\\xb8f\\xa9@\\xfeT\\x17\\xee\\xb9$\\\"\\xa2P\\x83\\xb0\\x94\\xf1\\xfc\\x92\\x1f\\x8f\\x7fX\\x91\\x14^\\xaa\\x02\\x05\\v\\\"(q$)\"][msg_match_2 \"Invalid character in request (null character)\"] [data_match_2 \"ARGS:C$V\\xc4\\x10\\x83qn0\\x95\\xe2\\x1d\\xa08\\xd9}D\\x99\\x9c\\x81S\\xa1i\\xf2\\xbd\\xb7RT\\x9f\\xbb\\xa7\\xef\\xaaN\\xb0ǖ\\xad\\xaa\\xc2mB\\xdd6\\xfb\\x99תd\\xcb\\x10\\xc2\\x16qhH\\xb1M\\xb9l_'^e\\xa1\\\\\\x98BO\\xb9Xn\\x95 \\x87KاNg\\xd7Q.\\x9b\\xcb\\xf2\\x93\\x81F\\u05ff\\xe6ʇhټ\\xe3fY2j\\xac\\xa1\\xa79\\xb7n\\x91\\xa8:p\\xd2T\\x85\\x19`\\x9b\\xb2\\xee\\xfa\\x9clRgy\\xe9E8\\xa90F\\x0fނj\\x98\\xef\\xfb\\xa4i\\xe6'\\x87e\\x86\\xc7\\xd3Sz~\\f\\xa2$\\xaaY\\x14\\xb5\\x91\\xef\\x92U!J$\\xea}\\xea\\x8d\\x04\\xeb\\xc6\\xe4\\xe5\\xd4pS)\\x1eTq\\
```

There is a bunch of rules being triggered, such as 941310, 941160, 921200, 941100, 942550, 933210, 933180, 933100, 932380, 932370, 932250, 932270, 932130, 932125, 932230, 932330, 920270, [...]. Is this related to enforce body processor in crs-setup.conf as you suggested, or missing entries in [this](https://github.com/coreruleset/nextcloud-rule-exclusions-plugin/blob/5397c3c441888dd49a29ade017ca5aaf1cff00e3/plugins/nextcloud-rule-exclusions-before.conf#L134) exclusion rule, or an upstream issue with coraza+caddy?

**Comment by EsadCetiner** [comment]
@ti-guru Don't edit the plugin files, your changes will be undone the next time you update the plugin. I recommend creating your own config file and then including that in Coraza:

```
              coraza_waf {
                      load_owasp_crs
                      directives `
                              Include /nextcloud-rule-exclusions-plugin/plugins/*.conf
                              Include /path/to/your-custom-rule-exclusions-file.conf
                              Include @coraza.conf-recommended
                              Include @crs-setup.conf.example
                              Include @owasp_crs/*.conf

                              #SecRuleRemoveById 921150
                              SecRequestBodyLimit 1073741824

                              SecRuleEngine On
                      `
              }
```

> Looking at the [crs-setup.conf](https://github.com/corazawaf/coraza-coreruleset/blob/9b73b5e90a09613c7535d391a15cc9cb08c05a8d/rules/%40crs-setup.conf.example#L217) it does not appear to be enabled.

hm... this false positive shouldn't be happening, can you share the full request from your audit log? Look for `VnlVOaXIbllaCyMR`

> Is this related to enforce body processor in crs-setup.conf as you suggested

Most likely, but it could be a result of an engine behavior difference between Coraza and ModSecurity. This plugin is only tested against Apache and NGINX with ModSecurity and not Coraza with Caddy. If the issue is an engine difference, then there are likely many other false positives that are undiscovered. If you could share all your related configs then that would help me narrow down the potential causes.

**Comment by EsadCetiner** [comment]
@ti-guru After some digging I found out that Coraza enables the `URLENCODED` body processor when no other body processor is configured (This is defined by the `modsecurity.conf` and `coraza.conf` recommended file), which is essentially the same as enabling `enforce body processor` in crs-setup.conf.

Can you try adding this rule to your Coraza config and see if this fixes your issue:
```
SecRule REQBODY_PROCESSOR "!@rx (?:URLENCODED|MULTIPART|XML|JSON)" \
    "id:105,\
    phase:1,\
    pass,\
    nolog,\
    ctl:requestBodyProcessor=RAW"
```

This will make Coraza behave like ModSecurity with Apache or NGINX.

**Comment by ti-guru** [comment]
@EsadCetiner Great catch and thank you for taking your time looking into this! This solved all errors described above, i.e. there are no errors when updating through DAVx5 and no error when uploading images. Should this be added to the troubleshooting in the readme?

**Comment by EsadCetiner** [comment]
@ti-guru Good to hear that the issue is resolved, feel free to open another issue if your still having problems.

> Should this be added to the troubleshooting in the readme?

Yes this will be added to the readme so others can be aware of this issue.
